backend/orcid.py

```diff
@@ -197,15 +214,21 @@ def fetch_orcid_records(self, orcid_identifier, profile=None, use_doi=True):
         # 1st attempt with DOIs and CrossRef
         if use_doi:
             # Let's grab papers with DOIs found in our ORCiD profile.
-            dois = [doi for doi, put_code in dois_and_putcodes]
-            for idx, (success, paper_or_metadata) in enumerate(self.fetch_metadata_from_dois(cr_api, ref_name, orcid_id, dois)):
+            dois = [doi for doi, _ in dois_and_putcodes]
+            metadata = self.fetch_metadata_from_dois(
+                cr_api, ref_name, orcid_id, dois
+            )
+            for idx, (success, paper_or_metadata) in enumerate(metadata):
                 if success:
-                    yield paper_or_metadata # We know that this is a paper
+                    yield paper_or_metadata  # We know that this is a paper
                 else:
                     put_codes.append(dois_and_putcodes[idx][1])
 
         # 2nd attempt with ORCID's own crappy metadata
-        works = profile.fetch_works(put_codes)
+        works = profile.fetch_works(
```

Comment:
Question rising from developper meeting: When we ingest the ORCID dump, we want to do it offline so we don't take into account the DOIs and the whole previous lines are skipped. We only use ORCID metadata (no DOI) which are incomplete.

So, ORCID dump should probably be imported **after** Crossref import is done and up to date. We could lookup for papers by DOIs (without queries to Crossref) if we have DOIs as well.

Any idea on the strategy here?

Reply:
I feel like we need to rethink the general architecture to handle this in a satisfactory way… If we can rely on the fact that Crossref is already imported, then ORCID should not be fetching its metadata a second time from the API, but only adding the ORCID information to the authors of the papers (adding an OaiRecord to keep track of the provenance).

But that means going away from the architecture where each paper importer emits a pure Python object (not linked to the database) representing all the paper metadata, and then a generic backend code deals with the saving to the DB. We currently don't follow that infrastructure completely, but that is the general direction I have been pushing towards.

Brainstorming session required!